sherpa/cpp_api/brpc/server.cc

```diff
@@ -0,0 +1,378 @@
+      }
+      return result_queue_[stream_id].data();
+    }
+    return nullptr;
+  }
+
+  int32_t segment_stream(const int64_t stream_id) {
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.find(stream_id) != stream_queue_.end()) {
+      return stream_queue_[stream_id]->IsEndpoint();
+    }
+    return false;
+  }
+
+  void batch_decode() {
+    while (alive_) {
+      {
+        std::unique_lock mlock(mutex_);
+        batch_cond_.wait_for(mlock, wait_ms_ * 1ms,
+            [this] { return batch_size_ <= stream_queue_.size(); });
+
+        std::vector<sherpa::OnlineStream *> ready_streams;
+        for (const auto & stream : stream_queue_) {
+          if (online_asr_.IsReady(stream.second.get())) {
+            ready_streams.push_back(stream.second.get());
+          }
+        }
+
+        if (ready_streams.empty()) { continue; }
+        decoding_ = true;
+        online_asr_.DecodeStreams(ready_streams.data(), ready_streams.size());
+        decoding_ = false;
+      }
+      decoding_cond_.notify_all();
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    exit_ = true;
+  }
+
+  void stop_decode() {
+    std::unique_lock mlock(mutex_);
+    alive_ = false;
+  }
+
+  void warmup_decode() {
+    auto s = online_asr_.CreateStream();
+    auto wav_tensor = torch::rand({32000}).to(torch::kFloat);
+    s->AcceptWaveform(SAMPLE_RATE, wav_tensor);
+  }
+
+ private:
+  mutable std::shared_mutex mutex_;
+
+  std::condition_variable_any batch_cond_;
+  std::condition_variable_any decoding_cond_;
+
+  bool alive_;
+  bool exit_;
+  bool decoding_;
+  const int32_t wait_ms_;
+  std::thread decoder_;
+
+  size_t batch_size_;
+  const size_t max_batch_size_;
+  sherpa::OnlineAsr online_asr_;
+  std::unordered_map<int64_t,
+    std::unique_ptr<sherpa::OnlineStream>> stream_queue_;
+  std::unordered_map<int64_t, std::vector<char>> result_queue_;
+  std::unordered_map<int64_t, int32_t> vad_state_queue_;
+};
+
+class StreamingAsrService : public sherpa::AsrService {
+ public:
+  explicit StreamingAsrService(const sherpa::OnlineAsrOptions opts,
+      const int32_t max_batch_size = 4096,
+      const int32_t wait_ms = 10)
+    : decoder_(std::make_shared<AsrDecoder>(opts, max_batch_size, wait_ms)),
+    tail_padding_(SAMPLE_RATE * sizeof(int16_t) * PADDING_LEN, 0) {}
+
+  ~StreamingAsrService() {
+    decoder_->stop_decode();
+  }
+
+  virtual void Recognize(google::protobuf::RpcController* controller,
+      const sherpa::AsrRequest * request,
+      sherpa::AsrResponse* response,
+      google::protobuf::Closure* done) {
+    try {
+      sherpa::AsrResponse_Status status = sherpa::AsrResponse::ok;
+      brpc::ClosureGuard done_guard(done);
+      brpc::Controller* cntl = static_cast<brpc::Controller*>(controller);
+      std::string remote_name(endpoint2str(cntl->remote_side()).c_str());
+      if (remote_id_.find(remote_name) == remote_id_.end()) {
+        std::unique_lock<std::mutex> mlock(mutex_);
```

Comment:
```suggestion
        std::lock_guard<std::mutex> mlock(mutex_);
```

`std::lock_guard` is more lightweight.